civic-tech-club/src/components/BudgetVis.js

```diff
@@ -0,0 +1,226 @@
+  'rgba(75, 192, 192, 1)',
+  'rgba(255, 99, 132, 1)',
+  'rgba(54, 162, 235, 1)',
+  'rgba(255, 206, 86, 1)',
+  'rgba(153, 102, 255, 1)',
+  'rgba(255, 159, 64, 1)',
+];
+
+const BudgetVis = () => {
+  const [chartTitle, setChartTitle] = useState('');
+  const [chartData, setChartData] = useState({ labels: [], datasets: [] });
+  const chartRef = useRef(null); // Create a ref for the chart
+
+  useEffect(() => {
+    // This code will run once when the component loads
+    handleChartChange('Gross Operating Expenditures');
+  }, []); // The empty array makes it run only once, when the component mounts
+
+  const handleChartChange = (title) => {
+    setChartTitle(title); // Set chart title for debugging or displaying
+    const newChartData = getLine(title, data); // Pass title directly to getLine function
+    setChartData(newChartData); // Update chart data with the new dataset
+    resetZoom();
+  };
+
+  // Generate the line chart data with custom colors
+  function getLine(budgetIndex, data) {
+    const chartdata = {
+      labels: ['2024', '2025', '2026', '2027'], // Preset labels for the X-axis (years)
+      datasets: [],
+    };
+
+    data.forEach((service, index) => {
+      let proceed = false;
+      let budgetObject = {};
+
+      service.budget.forEach((element) => {
+        if (element['Budget Summary'] === budgetIndex) {
+          proceed = true;
+          budgetObject = element;
+        }
+      });
+
+      if (proceed) {
+        // Create a dataset for each service
+        const indivObject = {
+          label: service['service'], // Dataset label
+          data: [
+            parseBudgetValue(budgetObject['2024']),
+            parseBudgetValue(budgetObject['2025']),
+            parseBudgetValue(budgetObject['2026']),
+            parseBudgetValue(budgetObject['2027']),
+          ],
+          borderColor: COLORS[index % COLORS.length], // Assign a unique color
+          fill: false, // No fill for line chart
+          tension: 0.1, // Smooth the line
+        };
+        chartdata.datasets.push(indivObject);
+      }
+    });
+
+    return chartdata;
+  }
+
+  // Helper function to clean up budget values
+  function parseBudgetValue(value) {
+    if (value === 'N/A' || value === '-$0') return null;
+    if (typeof value === 'string') {
+      value = value.replace(/[\$,]/g, '').trim(); // Remove dollar signs, commas, and spaces
+      return parseFloat(value);
+    }
+    return value;
+  }
+
+  // Custom plugin to place labels at the end of each line
+  const endLabelPlugin = {
+    id: 'endLabel',
+    afterDatasetDraw(chart, args, options) {
+      const { ctx } = chart;
+      chart.data.datasets.forEach((dataset, i) => {
+        const meta = chart.getDatasetMeta(i);
+        const lastPoint = meta.data[meta.data.length - 1]; // Get last point of each line
+        const x = lastPoint.x +10; // Offset label a bit to the right of the point
+        const y = lastPoint.y;
+
+        ctx.save();
```

Comment:
do we need to manage these calls ourselves? Doesn't the charting library handle this?

Reply:
The charting library seems to only have the option of using a legend, which doesn't really work for us as there are so many different lines. The only way I could figure out how to do it is pretty manual. It's not perfect either, it gets cut off if the label is too long. If you've got any ideas I'm open to hearing them

Reply:
ok good to know!